source/bouncy_castle.f90

```diff
@@ -756,37 +761,33 @@ real(kind=real64) function asin_rd(x)
   end function asin_rd
 
   real(kind=real64) function atan2_rd(y,x)
-    use, intrinsic :: iso_fortran_env
-    use, intrinsic :: ieee_arithmetic
     implicit none
     real(kind=real64) x,y
-    logical myisnan
+    if (x.eq.0d0) then
+      if (y.eq.0d0) then
 #ifdef NAGFOR
-    real(kind=real64) :: nan64
-    nan64 = ieee_value(nan64, ieee_quiet_nan)
+        atan2_rd=0d0
 #else
-    real(kind=real64), parameter :: nan64 = transfer(-2251799813685248_int64, 1.0_real64)
+        ! Let the internal atan2 handle this case according to ieee
+        atan2_rd=atan2(y,x)
 #endif
-    if (x.eq.0d0) then
-       if (y.eq.0d0) then
-          ! Should get me a NaN
-          atan2_rd=atan_rd(y/x)
-       else
-          atan2_rd=sign(mb_pi2,y)
-       endif
+      else
+        atan2_rd=sign(mb_pi2,y)
+      endif
     else
-       if (y.eq.0d0) then
-          if (x.gt.0d0) then
-             atan2_rd=0d0
-          else
-             atan2_rd=mb_pi
-          endif
-       else
-          atan2_rd=atan_rd(y/x)
-          if (x.lt.0d0) then
-             atan2_rd=sign(mb_pi,y)+atan2_rd
-          endif
-       endif
+      if (y.eq.0d0) then
```

Comment:
Could this if/else logic be moved to the calling `_mb` function, instead of being duplicated?

Reply:
Yes, there are several things that can be improved in the flow of this module. I don't have the time for re-structuring the code now.